web/elm/src/Routes.elm

```diff
@@ -188,9 +190,34 @@ resource =
             <?> Query.int "from"
             <?> Query.int "to"
             <?> Query.int "limit"
+            <?> resourceVersion "filter"
         )
 
 
+resourceVersion : String -> Query.Parser (Maybe Concourse.Version)
+resourceVersion key =
+    let
+        split s =
+            case String.split ":" s of
+                [ k, v ] ->
+                    Just ( k, v )
```

Comment:
We'll need to rethink this a bit, as it doesn't work if the key/value contains a `:` (which happens a lot, actually, since the `digest` of the `registry-image` resource has a `:`) - if I click on the version of a resource using `registry-image`, it just strips out the `digest` from the URL and only keeps the `tag`.

The easiest solution would be to just encode a full JSON object and throw that in a single `filter` query param.

Another fairly easy solution is for `Concourse.versionQuery` to `URL.percentEncode` the key and value separately, and `URL.percentDecode` them separately here as well. That way, they end up getting doubly-encoded. A similar approach is to base64 encode them, but that makes the query string a little opaque.

If you want to avoid an ugly query string (because of all the special characters that need escaping in the other approaches), you could do something like I did with instance vars (`vars.key1="value1"&vars.key2="value2"`) - the serialization/parsing is all in https://github.com/concourse/concourse/blob/master/web/elm/src/DotNotation.elm. Since it's just a `Dict String String` it's even simpler - you don't need to worry about recursing through JSON objects or quoting strings with special characters. However, this approach requires a bit of a hack since the query parameter names are dynamic:
https://github.com/concourse/concourse/blob/ecda93e38208d0971e280076f9fbe8864d8e90dc/web/elm/src/Routes.elm#L521-L532

I personally think the JSON approach is the easiest and probably the best - I tried to make the instance vars query params more friendly since they appear all the time, but in this case you probably won't need to see it that often.

Reply:
I originally had it in the JSON approach, but I took that out to make it more similar to the atc api logic. My thought was that if it fails here, then it'll also fail when the api server tries to decode it. What do you think of just taking the head of the list as the key and concatenate the rest into a string for the value?

https://github.com/concourse/concourse/blob/fc2fda3221c19f7ba39b2e9b96fa492463129e34/atc/api/resourceserver/versionserver/list.go#L34-L41

Reply:
Huh, wasn't aware of that! seems a bit problematic in general that the API server doesn't allow for keys that have a `:`, but oh well. I think your suggestion of taking the head as the key and the tail (joined by `:`) as the value is reasonable, since that'd match the behaviour of `SplitN(field, ":", 2)`